File: Chapter 2/resources/java/src/sort/InsertionSort.java
package sort;

import search.BinarySearch;

import java.util.Arrays;

public class InsertionSort {
    protected static int[] insertionSort(int[] array) {
        for (int i = 1; i < array.length; i++) {
            int currentValue = array[i];

            int checkIndex = i - 1;
            while (checkIndex >= 0 && array[checkIndex] > currentValue) {
                array[checkIndex + 1] = array[checkIndex];
                checkIndex -= 1;
            }
            array[checkIndex + 1] = currentValue;
        }
        return array;
    }

    protected static int[] reverseInsertionSort(int[] array) {
        for (int i = 1; i < array.length; i++) {
            int currentValue = array[i];

            int checkIndex = i - 1;
            while (checkIndex >= 0 && array[checkIndex] < currentValue) {
                array[checkIndex + 1] = array[checkIndex];
                checkIndex -= 1;
            }
            array[checkIndex + 1] = currentValue;
        }
        return array;
    }

    protected static void recInsertionSort(int[] array) {
        recInsertionSort(array, array.length);
    }

    private static void recInsertionSort(int[] array, int n) {
        if (n <= 1)
            return;

        recInsertionSort(array, n - 1);

        int current = array[n - 1];
        int i = n - 2;

        while (i >= 0 && array[i] > current) {
            array[i + 1] = array[i];
            i--;
        }
        array[i + 1] = current;
    }

    protected static int[] insertionSortWithBinarySearch(int[] array) {
        for (int i = 1; i < array.length; i++) {
            int currentValue = array[i];
            int pos = BinarySearch.recSearchPosition(array, 0, i-1, currentValue);

            for (int j = i; j > pos; j--) {
                array[j] = array[j - 1];
            }
            array[pos] = currentValue;
        }
        return array;
    }
}

File: Chapter 2/resources/rust/src/sort/insertion_sort.rs
use crate::search::binary_search::rec_find_position;

fn insertion_sort(array: &mut [i32]) {
    for i in 1..array.len() {
        let mut current_index = i;
        while current_index > 0 && array[current_index] < array[current_index - 1] {
            array.swap(current_index, current_index - 1);
            current_index -= 1;
        }
    }
}

fn reverse_insertion_sort(array: &mut [i32]) {
    for i in 1..array.len() {
        let mut current_index = i;
        while current_index > 0 && array[current_index] > array[current_index - 1] {
            array.swap(current_index, current_index - 1);
            current_index -= 1;
        }
    }
}

fn rec_insertion_sort(array: &mut [i32], n: usize) {
    if n <= 1 {
        return;
    }

    rec_insertion_sort(array, n - 1);

    let current = array[n - 1];
    let mut i = n - 2;
    let mut flag = false;

    while !flag && array[i] > current {
        array[i + 1] = array[i];
        if i > 0 {
            i -= 1;
        } else {
            flag = true
        }
    }
    if !flag {
        i += 1
    }

    array[i] = current
}

fn insertion_sort_with_binary_search(array: &mut [i32]) {
    for i in 1..array.len() {
        let current_value = array[i];
        let pos = rec_find_position(array,0, i-1,current_value);

        for j in (pos..i).rev() {
            array.swap(j, j+1);
        }
        array[pos] = current_value
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn insertion_sort() {
        let mut array = [31, 41, 59, 26, 41, 58];
        super::insertion_sort(&mut array);
        assert_eq!([26, 31, 41, 41, 58, 59], array);
    }

    #[test]
    fn reverse_insertion_sort() {
        let mut array = [31, 41, 59, 26, 41, 58];
        super::reverse_insertion_sort(&mut array);
        assert_eq!([59, 58, 41, 41, 31, 26], array);
    }

    #[test]
    fn rec_insertion_sort() {
        let mut array = [31, 41, 59, 26, 41, 58];
        let len = array.len();
        println!("{} : {:?}", 0, array);
        super::rec_insertion_sort(&mut array, len);
        assert_eq!([26, 31, 41, 41, 58, 59], array);
    }

    #[test]
    fn insertion_sort_with_binary_search() {
        let mut array = [31, 41, 59, 26, 41, 58];
        super::insertion_sort_with_binary_search(&mut array);
        assert_eq!([26, 31, 41, 41, 58, 59], array);
    }
}
